Run InferCNV for Ewing's sample (#432)
AlexsLemonade/OpenScPCA-analysis: analyses/cell-type-ewings/exploratory_analysis/04-infercnv.Rmd

- `has_cnv_chr1`
- `has_loss_chr1`
- `has_dupli_chr1`
- `proportion_cnv_chr1`
- `proportion_loss_chr1`
- `proportion_dupli_chr1`
- `proportion_scaled_cnv_chr1`
- `proportion_scaled_loss_chr1`
- `proportion_scaled_dupli_chr1`

The `has` columns are boolean columns indicating if that cell has presence of a CNV, loss, or duplication event. 
If `has_cnv` is TRUE if either `has_loss` or `has_dupli` is TRUE. 

See [the comments in this issue](https://github.com/broadinstitute/infercnv/issues/208) explaining what the contents of the columns are. 

One approach to classifying tumor cells is to look for cells that have a lot of CNVs. 
Let's start by just summing up the total number of chromosomes with CNVs per cell and plotting that. 

```{r}
# grab any columns that have has_cnv 
has_cnv_cols <- names(all_cnv_df)[startsWith(colnames(all_cnv_df), "has_cnv")]

# don't include the normal cells that were provided as reference in scaling 
cnv_df <- all_cnv_df |>
  # get total number of chromosomes with cnvs per cell and then scale that number 
  dplyr::mutate(
    has_cnv = ifelse(
      stringr::str_detect(subcluster, "reference"), 
      NA_real_, 
      rowSums(across(has_cnv_cols))
    ),

>     z_scaled_cnv = scale(has_cnv))

Reviewer: I was not super comfortable with this before, and on a bit more reflection, I think we want to just remove it. The reason is that the values of `has_cnv` are not continuous, but rather in the range [0-22], and that is not going to change. So I think we probably want to just use that knowledge straight. So below, I would say you should use histograms rather than density plots.

```

```{r}
# plot the total number of CNVs
ggplot(cnv_df, aes(x = UMAP1, y= UMAP2, color = has_cnv)) +
  geom_point(alpha = 0.5, size = 0.5) +
  theme_bw() +
  scale_color_viridis_c()
```
This shows that the total CNVs are highest in the group of cells that correspond to the same cells that are manually annotated as tumor cells, which is promising. 
Those cells in grey (bottom left) are cells used as a reference. 
If we were to use total CNVs we would need some sort of cut off for the number. 
Let's look at the distribution of total CNVs and also the scaled total CNVs to see if we can identify a cut off. 

```{r}
ggplot(cnv_df, aes(x = has_cnv)) +

>   geom_density() +

Reviewer: Try a histogram here instead? (binwidth and center to make things line up over the integers)
```suggestion
  geom_histogram(bindwith = 1, center = 0, alpha = 0.5, position="identity") +
```

alternatively: 
```suggestion
  geom_freqpoly(binwidth = 1, center = 0) +
```

  theme_bw() +
  labs(x = "Total number of CNVs")

ggplot(cnv_df, aes(x= z_scaled_cnv)) + 
  geom_density() +
  theme_bw() +
  labs(x = "z-scaled chromosome CNV count")
```

> Looking at the scaled number of CNVs it does look like we could use 0 as the cutoff and call any cell that has > 0 CNVs (after scaling) as tumor and all other cells as normal. 

Reviewer: This is the equivalent of just using the mean as the cutoff, so let's make that explicit?

In the below plot we label all non-reference cells as either tumor or normal. 
Cells that are NA are those that were already classified as reference cells. 

```{r}
# add cnv classification 
cnv_df <- cnv_df |> 

>   dplyr::mutate(cnv_classification = dplyr::if_else(z_scaled_cnv > 0, "Tumor", "Normal"))

Reviewer: Explicitly using the mean.
```suggestion
  dplyr::mutate(cnv_classification = dplyr::if_else(has_cnv > mean(has_cnv), "Tumor", "Normal"))
```

  geom_point(alpha = 0.5, size = 0.5) +
  theme_bw() 
```

Another option is to use the scaled proportion of CNVs which tells us the proportion of genes on that chromosome that are part of the CNV.  

```{r}
proportion_cols <- names(cnv_df)[startsWith(colnames(cnv_df), "proportion_scaled_cnv")]

# get total number of genes per chromosome to use in weighted mean 
chr_weights <- final_cnv_obj@gene_order |> 
  as.data.frame() |> 
  dplyr::count(chr) |> 
  # only keep chr 1-22, drops MT and GL chr
  dplyr::filter(chr %in% glue::glue("chr{seq(1,22)}")) |> 
  dplyr::pull(n)

# get mean proportion of cnvs per cell weighted by chromsome size (number of genes)
cnv_df <- cnv_df |> 
  dplyr::rowwise() |> 
  # calculate weighted mean, excluding normal cells 
  dplyr::mutate(
    mean_proportion = ifelse(
      !is.na(has_cnv), 
      weighted.mean(across(proportion_cols), chr_weights), 
      NA)) |> 
  dplyr::ungroup()
```

```{r}
ggplot(cnv_df, aes(x = UMAP1, y= UMAP2, color = mean_proportion)) +
  geom_point(alpha = 0.5, size = 0.5) +
  theme_bw() +
  scale_color_viridis_c()
```

```{r}
ggplot(cnv_df, aes(x = mean_proportion)) +
  geom_density() +
  theme_bw()
```

> It looks like we could use a cutoff of 0.1 here to classify tumor/normal cells. 

Reviewer: I might also look at using the mean here? This will be a more productive discussion with more samples, of course.

Author: We are taking the weighted mean here so want to check that you just want to use the mean alone? Or compare calling tumor cells using the weighted mean to the mean of the CNV sum?

Reviewer: I think yes. Specifically, the cutoff I am proposing is `mean(cnv_df$mean_proportion)`

Author: Sorry I was confused and thought you meant to look at the distribution of the mean and wasn't clear that you meant to use it as the cutoff. But that makes sense so I'm updating it to use that!

- Gain of Chr8
- Gain of Chr12
- Gain of Chr1p
- Loss of Chr16q

Although these are the most frequent, there are patients who do not have any of these alterations and patients that only have some of these alterations. 
See [Tirode et al.,](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4264969/) and [Crompton et al.](https://doi.org/10.1158/2159-8290.CD-13-1037). 

First we will compare the proportions across each chromosome between cells we have labeled as tumor or normal. 
Then we will look specifically at the chromosomes commonly affected in Ewing sarcoma. 

```{r}
# prepare a data frame with one row per chromosome per cell for plotting
proportion_df <- cnv_df |> 
  # select only relevant columns for plotting
  dplyr::select(barcodes, UMAP1, UMAP2, cnv_mean_classification, all_of(proportion_cols)) |> 
  tidyr::pivot_longer(cols = all_of(proportion_cols),
    names_to = "chromosome",
    values_to = "proportion") |> 
  # remove extra string in front of chr 
  dplyr::mutate(
    chromosome = stringr::str_remove(chromosome, "proportion_scaled_cnv_chr"),
    # turn into numeric for easy ordering in plot 
    chromosome = as.numeric(chromosome)
  )

```

> ```{r, fig.height=10, fig.width=10}
> ggplot(proportion_df, aes(x = proportion, color = cnv_mean_classification)) +
>   geom_density() +
>   theme_bw() +
>   facet_wrap(vars(chromosome), ncol = 3)
> ```

Reviewer: A couple adjustments to the density plot here to make it more readable: 

One is to add bounds to the  density calculation, which should help make the plot not extend past (0,1), but you may still want to set the width of the plot as well with `coord_cartesian()`. I also set a bandwidth (which was just a guess, definitely see if you can adjust) to make the plots more comparable. The goal here is to make the plots less spikey than the default, with a secondary goal of making them more similar in the y-range. 

The other thing you could do is add scaling to the density plots so they all have a max of 1: `geom_density(aes(y = after_stat(scaled))`, which is not quite kosher, but does often make comparing shapes easier.

````suggestion
```{r, fig.height=10, fig.width=10}
ggplot(proportion_df, aes(x = proportion, color = cnv_mean_classification)) +
  geom_density(bounds = c(0,1), bw = 0.05) +
  theme_bw() +
  facet_wrap(vars(chromosome), ncol = 3)
```
````

# The path to this module
module_base <- file.path(repository_base, "analyses", "cell-type-ewings")
```

```{r}
# source in helper functions
jaccard_functions <- file.path(module_base, "scripts", "utils", "jaccard-functions.R")
source(jaccard_functions)
```

```{r paths}
# Input files
sce_filename <- glue::glue("{params$library_id}_processed.rds")
sce_file <- file.path(sample_dir, sce_filename)

# tumor normal classifications 
classifications_results_dir <- file.path(module_base, "results", "marker_gene_analysis")
classifications_filename <- glue::glue("{params$library_id}_tumor_normal_classifications.tsv")
classifications_file <- file.path(classifications_results_dir, classifications_filename)

# output from running infercnv 
results_dir <- file.path(module_base, "results", "infercnv", params$library_id)

all_results_dir_list <- list(
  no_ref = file.path(results_dir, "no_reference"),
  with_ref = file.path(results_dir, "endothelial_reference")
)

# png ref files 
png_files <- all_results_dir_list |>
  purrr::map_chr(\(dir) file.path(dir, glue::glue("{params$library_id}_infercnv.png")))

png_files |> 
  # make sure png files exists

>   purrr::map(\(file)

Reviewer: Prevent extra output
```suggestion
  purrr::walk(\(file)
```

The top group contains all cells that were used as the reference group and the bottom is all other cells. 
The color indicates the relative expression intensity of each gene with the darker colors either indicating a gain of expression or loss of expression compared to the reference cells.

From the [`InferCNV` documentation](https://github.com/broadinstitute/inferCNV/wiki/Interpreting-the-figure): 

> The normal cells in the top heatmap define baseline expression for genes in normal cells. This baseline distribution of normal gene expression is subtracted from both the normal cells from which was defined as well as the tumor cells. Afterwards, the normal cell expression heatmap should be largely devoid of signal, with the exception of certain outlier gene expression values in certain cells. Removing this baseline of normal expression signal from the tumor cells should reveal those chromosomal regions that have significantly more or less expression than the normal cells, highlighting likely amplified or deleted whole chromosomes or large chromosomal regions.

The dendogram on the left side is hierarchical clustering of all non-reference cells. 

> The color annotation next to it indicates subdivisions of the dendogram by cutting it in `k_obs_groups` groups, with the default `k_obs_groups` = 1. 

Reviewer: Which color annotation is this? There are two, it seems to me, and I can't tell what the rainbow is: it certainly isn't `k_obs_groups = 1`!

Author: I'm having trouble understanding the docs here, because it sounds like the bar immediately next to the dendogram (the rainbow one) is the subdivisions of the dendogram. I think there is an error in the docs, because you are correct that is not `k_obs_groups=1`. And the right most bar is supposed to be the original annotation provided.

chr_columns_to_keep <- which(!stringr::str_detect(names(coldata_df), "chrMT|chrGL000219.1|chrGL000194.1"))
all_cnv_df <- coldata_df[, chr_columns_to_keep]
```

The metadata that we added includes the following columns for each chromosome: 

- `has_cnv_chr1`
- `has_loss_chr1`
- `has_dupli_chr1`
- `proportion_cnv_chr1`
- `proportion_loss_chr1`
- `proportion_dupli_chr1`
- `proportion_scaled_cnv_chr1`
- `proportion_scaled_loss_chr1`
- `proportion_scaled_dupli_chr1`

The `has` columns are boolean columns indicating if that cell has presence of a CNV, loss, or duplication event. 
If `has_cnv` is TRUE if either `has_loss` or `has_dupli` is TRUE. 

See [the comments in this issue](https://github.com/broadinstitute/infercnv/issues/208) explaining what the contents of the columns are. 

One approach to classifying tumor cells is to look for cells that have a lot of CNVs. 
Let's start by just summing up the total number of chromosomes with CNVs per cell and plotting that. 

```{r}
# grab any columns that have has_cnv 
has_cnv_cols <- names(all_cnv_df)[startsWith(colnames(all_cnv_df), "has_cnv")]

> # don't include the normal cells that were provided as reference in scaling 
> reference_cells_only <- all_cnv_df |> 
>   dplyr::filter(stringr::str_detect(subcluster, "reference")) |> 
>   dplyr::mutate(has_cnv = NA,
>                 z_scaled_cnv = NA)
> 
> unknown_cells_only <- all_cnv_df |>
>   dplyr::filter(stringr::str_detect(subcluster, "unknown")) |> 
>   # get total number of chromosomes with cnvs per cell and then scale that number 
>   dplyr::mutate(has_cnv = rowSums(across(has_cnv_cols)),
>                 z_scaled_cnv = scale(has_cnv))

Reviewer: Rather than splitting these up and then rejoining, I would probably do something like this:
```suggestion
# don't include the normal cells that were provided as reference in scaling 
cnv_df <- all_cnv_df |>
  # get total number of chromosomes with cnvs per cell and then scale that number 
  dplyr::mutate(
    has_cnv = ifelse(
      stringr::str_detect(subcluster, "reference"), 
      NA_real_, 
      rowSums(across(has_cnv_cols))
    ),
    z_scaled_cnv = scale(has_cnv))
```

`scale` unlike `mean` and `sum` seems to handle `NA` values as you would want it to.

ggplot(cnv_df, aes(x= z_scaled_cnv)) + 
  geom_density() +
  theme_bw() +
  labs(x = "z-scaled chromosome CNV count")
```

Looking at the scaled number of CNVs it does look like we could use 0 as the cutoff and call any cell that has > 0 CNVs (after scaling) as tumor and all other cells as normal. 

In the below plot we label all non-reference cells as either tumor or normal. 
Cells that are NA are those that were already classified as reference cells. 

```{r}
# add cnv classification 
cnv_df <- cnv_df |> 
  dplyr::mutate(cnv_classification = dplyr::if_else(z_scaled_cnv > 0, "Tumor", "Normal"))

ggplot(cnv_df, aes(x = UMAP1, y= UMAP2, color = cnv_classification)) +
  geom_point(alpha = 0.5, size = 0.5) +
  theme_bw() 
```

Another option is to use the scaled proportion of CNVs which tells us the proportion of genes on that chromosome that are part of the CNV.  

```{r}
proportion_cols <- names(cnv_df)[startsWith(colnames(cnv_df), "proportion_scaled_cnv")]

# get total number of genes per chromosome to use in weighted mean 
chr_weights <- final_cnv_obj@gene_order |> 
  as.data.frame() |> 
  dplyr::count(chr) |> 
  # only keep chr 1-22, drops MT and GL chr

>   dplyr::filter(chr %in% glue::glue("chr{seq(1,22)}")) |> 

Reviewer: Do you also want X?

Author: ChrX isn't in the output as an option.

  dplyr::pull(n)

# get mean proportion of cnvs per cell weighted by chromsome size (number of genes)
cnv_df <- cnv_df |> 
  dplyr::rowwise() |> 
  # calculate weighted mean, excluding normal cells 

>   dplyr::mutate(mean_proportion = ifelse(!is.na(has_cnv), weighted.mean(across(proportion_cols), num_genes_df$n), NA)) |> 

Reviewer: You did it in one step this time, but I would break up across lines for clarity

Reviewer: You are using `num_genes_df` here, but I think this should be `chr_weights` after the code? I also think this is probably not quite right, as the MT genes have not been excluded from the proportion columns.

Rather than filtering out those chromosomes above, I might just set their weights to 0 so you don't have to worry about mismatched lengths?

Author: Sorry the `num_genes_df` is leftover from when that's what it was named. Also I have removed the MT genes from all columns so the dimensions should match up correctly.

  dplyr::ungroup()
```

```{r}
ggplot(cnv_df, aes(x = UMAP1, y= UMAP2, color = mean_proportion)) +
  geom_point(alpha = 0.5, size = 0.5) +
  theme_bw() +
  scale_color_viridis_c()
```

```{r}
ggplot(cnv_df, aes(x = mean_proportion)) +
  geom_density() +
  theme_bw()
```

> It looks like we could use a cutoff of 0.1 here to classify tumor/normal cells. 

Reviewer: For formatting
```suggestion

It looks like we could use a cutoff of 0.1 here to classify tumor/normal cells. 
```
